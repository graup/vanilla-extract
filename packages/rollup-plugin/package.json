{
  "name": "@vanilla-extract/rollup-plugin",
  "version": "1.0.0",
  "description": "Zero-runtime Stylesheets-in-TypeScript",
  "main": "dist/vanilla-extract-rollup-plugin.cjs.js",
  "module": "dist/vanilla-extract-rollup-plugin.esm.js",
  "files": [
    "/dist"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/seek-oss/vanilla-extract.git",
    "directory": "packages/rollup-plugin"
  },
  "author": "SEEK",
  "license": "MIT",
  "dependencies": {
    "@vanilla-extract/integration": "^3.0.0"
  },
  "devDependencies": {
    "rollup": "^2.0.0"
  },
  "peerDependencies": {
    "rollup": "^2.0.0"
  }
}
