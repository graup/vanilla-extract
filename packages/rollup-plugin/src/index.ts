import type { Plugin } from 'rollup';
import {
  cssFileFilter,
  processVanillaFile,
  compile,
  IdentifierOption,
  getSourceFromVirtualCssFile,
  virtualCssFileFilter,
} from '@vanilla-extract/integration';
import { relative, normalize, dirname } from 'path';

interface Options {
  identifiers?: IdentifierOption;
  cwd?: string;
}
export function vanillaExtractPlugin({ identifiers, cwd = process.cwd() }: Options = {}): Plugin {
  const emittedFiles = new Map<string, string>();
  const isProduction = process.env.NODE_ENV === 'production';

  return {
    name: 'vanilla-extract',
    async transform(_code, id) {
      if (!cssFileFilter.test(id)) {
        return null;
      }

      const index = id.indexOf('?');
      const filePath = index === -1 ? id : id.substring(0, index);

      const { source, watchFiles } = await compile({
        filePath,
        cwd,
      });

      for (const file of watchFiles) {
        this.addWatchFile(file);
      }

      return processVanillaFile({
        source,
        filePath,
        identOption: identifiers ?? (isProduction ? 'short' : 'debug'),
      });
    },
    async resolveId(id) {
      // Only handle virtual css modules here
      if (!virtualCssFileFilter.test(id)) {
        return null;
      }

      // Emit the css asset
      const { fileName, source } = await getSourceFromVirtualCssFile(id);
      if (!emittedFiles.get(fileName)) {
        const assetId = this.emitFile({
          type: 'asset',
          name: fileName,
          source,
        });
        emittedFiles.set(fileName, assetId);
      }

      // Resolve to an external chunk
      return {
        id: fileName,
        external: true,
      };
    },
    renderChunk(code, chunkInfo) {
      // For all imports that we emitted files for...
      const importsToReplace = chunkInfo.imports.filter((fileName) => emittedFiles.get(fileName));
      if (!importsToReplace.length) return null;

      // Replace css imports in chunks with relative paths to emitted css files
      const chunkPath = dirname(chunkInfo.fileName);
      return importsToReplace.reduce((codeResult, importPath) => {
        const assetId = emittedFiles.get(importPath)!;
        const assetName = this.getFileName(assetId);
        const fixedImportPath = `./${normalize(relative(chunkPath, assetName))}`;
        return codeResult.replace(importPath, fixedImportPath);
      }, code);
    },
  };
}
