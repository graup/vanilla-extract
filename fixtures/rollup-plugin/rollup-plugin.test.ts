import { rollup, OutputOptions } from 'rollup';
import { vanillaExtractPlugin } from '@vanilla-extract/rollup-plugin';
import esbuild from 'rollup-plugin-esbuild';
import { resolve } from 'path';

async function buildAndMatchSnapshot(outputOptions: OutputOptions) {
  const inputOptions = {
    input: resolve(__dirname, 'src/index.ts'),
    plugins: [vanillaExtractPlugin({ cwd: __dirname }), esbuild()],
  };
  const bundle = await rollup(inputOptions);
  const { output } = await bundle.generate(outputOptions);
  output.sort((a, b) => a.fileName.localeCompare(b.fileName));
  expect(output.map((chunkOrAsset) => ({
    name: chunkOrAsset.name,
    fileName: chunkOrAsset.fileName,
    code: chunkOrAsset.type === 'asset' ? chunkOrAsset.source : chunkOrAsset.code,
  }))).toMatchSnapshot();
  if (bundle) {
    await bundle.close();
  }
}

describe('rollup-plugin', () => {
  it('should build without preserveModules', async () => {
    // Bundle all JS outputs together
    const outputOptions = {
      format: 'esm',
    } as const;
    await buildAndMatchSnapshot(outputOptions);
  });

  it('should build with preserveModules', async () => {
    // Preserve JS modules
    const outputOptions = {
      format: 'esm',
      preserveModules: true,
    } as const;
    await buildAndMatchSnapshot(outputOptions);
  });
  
  it('should build with preserveModules and assetFileNames', async () => {
    // Preserve JS modules and place assets next to JS files instead of assets directory
    const outputOptions = {
      format: 'esm',
      preserveModules: true,
      assetFileNames({ name }) {
        return name.replace(new RegExp(`^@fixtures/rollup-plugin/src/`), '');
      },
    } as const;
    await buildAndMatchSnapshot(outputOptions);
  });
});

