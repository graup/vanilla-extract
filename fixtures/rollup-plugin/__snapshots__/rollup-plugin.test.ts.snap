// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollup-plugin should build with preserveModules 1`] = `
Array [
  Object {
    "code": ".Button__rscmzm0 {
  border-radius: 111px;
}",
    "fileName": "assets/@fixtures/rollup-plugin/src/Button.css.ts.vanilla-9d92f183.css",
    "name": "@fixtures/rollup-plugin/src/Button.css.ts.vanilla.css",
  },
  Object {
    "code": ".style__pbyrxo0 {
  border-radius: 999px;
}",
    "fileName": "assets/@fixtures/rollup-plugin/src/Button/style.css.ts.vanilla-e563dbfb.css",
    "name": "@fixtures/rollup-plugin/src/Button/style.css.ts.vanilla.css",
  },
  Object {
    "code": ".style__1352h0c0 {
  border-radius: 555px;
}",
    "fileName": "assets/@fixtures/rollup-plugin/src/style.css.ts.vanilla-4a3e9f3b.css",
    "name": "@fixtures/rollup-plugin/src/style.css.ts.vanilla.css",
  },
  Object {
    "code": "import './assets/@fixtures/rollup-plugin/src/Button.css.ts.vanilla-9d92f183.css';

var button1 = \\"Button__rscmzm0\\";

export { button1 };
",
    "fileName": "Button.css.js",
    "name": "Button.css",
  },
  Object {
    "code": "import './../assets/@fixtures/rollup-plugin/src/Button/style.css.ts.vanilla-e563dbfb.css';

var button3 = \\"style__pbyrxo0\\";

export { button3 };
",
    "fileName": "Button/style.css.js",
    "name": "style.css",
  },
  Object {
    "code": "export { button1 } from './Button.css.js';
export { button2 } from './style.css.js';
export { button3 } from './Button/style.css.js';
",
    "fileName": "index.js",
    "name": "index",
  },
  Object {
    "code": "import './assets/@fixtures/rollup-plugin/src/style.css.ts.vanilla-4a3e9f3b.css';

var button2 = \\"style__1352h0c0\\";

export { button2 };
",
    "fileName": "style.css.js",
    "name": "style.css",
  },
]
`;

exports[`rollup-plugin should build with preserveModules and assetFileNames 1`] = `
Array [
  Object {
    "code": "import './Button.css.ts.vanilla.css';

var button1 = \\"Button__rscmzm0\\";

export { button1 };
",
    "fileName": "Button.css.js",
    "name": "Button.css",
  },
  Object {
    "code": ".Button__rscmzm0 {
  border-radius: 111px;
}",
    "fileName": "Button.css.ts.vanilla.css",
    "name": "@fixtures/rollup-plugin/src/Button.css.ts.vanilla.css",
  },
  Object {
    "code": "import './style.css.ts.vanilla.css';

var button3 = \\"style__pbyrxo0\\";

export { button3 };
",
    "fileName": "Button/style.css.js",
    "name": "style.css",
  },
  Object {
    "code": ".style__pbyrxo0 {
  border-radius: 999px;
}",
    "fileName": "Button/style.css.ts.vanilla.css",
    "name": "@fixtures/rollup-plugin/src/Button/style.css.ts.vanilla.css",
  },
  Object {
    "code": "export { button1 } from './Button.css.js';
export { button2 } from './style.css.js';
export { button3 } from './Button/style.css.js';
",
    "fileName": "index.js",
    "name": "index",
  },
  Object {
    "code": "import './style.css.ts.vanilla.css';

var button2 = \\"style__1352h0c0\\";

export { button2 };
",
    "fileName": "style.css.js",
    "name": "style.css",
  },
  Object {
    "code": ".style__1352h0c0 {
  border-radius: 555px;
}",
    "fileName": "style.css.ts.vanilla.css",
    "name": "@fixtures/rollup-plugin/src/style.css.ts.vanilla.css",
  },
]
`;

exports[`rollup-plugin should build without preserveModules 1`] = `
Array [
  Object {
    "code": ".Button__rscmzm0 {
  border-radius: 111px;
}",
    "fileName": "assets/@fixtures/rollup-plugin/src/Button.css.ts.vanilla-9d92f183.css",
    "name": "@fixtures/rollup-plugin/src/Button.css.ts.vanilla.css",
  },
  Object {
    "code": ".style__pbyrxo0 {
  border-radius: 999px;
}",
    "fileName": "assets/@fixtures/rollup-plugin/src/Button/style.css.ts.vanilla-e563dbfb.css",
    "name": "@fixtures/rollup-plugin/src/Button/style.css.ts.vanilla.css",
  },
  Object {
    "code": ".style__1352h0c0 {
  border-radius: 555px;
}",
    "fileName": "assets/@fixtures/rollup-plugin/src/style.css.ts.vanilla-4a3e9f3b.css",
    "name": "@fixtures/rollup-plugin/src/style.css.ts.vanilla.css",
  },
  Object {
    "code": "import './assets/@fixtures/rollup-plugin/src/Button.css.ts.vanilla-9d92f183.css';
import './assets/@fixtures/rollup-plugin/src/style.css.ts.vanilla-4a3e9f3b.css';
import './assets/@fixtures/rollup-plugin/src/Button/style.css.ts.vanilla-e563dbfb.css';

var button1 = \\"Button__rscmzm0\\";

var button2 = \\"style__1352h0c0\\";

var button3 = \\"style__pbyrxo0\\";

export { button1, button2, button3 };
",
    "fileName": "index.js",
    "name": "index",
  },
]
`;
